File: app/channels/chat_room_channel.js
import consumer from "./consumer"

consumer.subscriptions.create({ channel: "ChatRoomChannel", chat_room_id: YOUR_CHAT_ROOM_ID }, {
  received(data) {
    // Called when there's incoming data on the websocket for this channel
    $('#messages').append(data['message']);
  }
});

$('#new_message').submit(function(e) {
  e.preventDefault();
  const messageContent = $('#message_content').val();
  consumer.send({ message: messageContent });
  $('#message_content').val('');
});

File: app/channels/chat_room_channel.rb
class ChatRoomChannel < ApplicationCable::Channel
  def subscribed
    stream_from "chat_room_#{params[:room_id]}"
  end

  def unsubscribed
    # Any cleanup needed when channel is unsubscribed
  end

  def send_message(data)
    message = Message.create(content: data['message'], user: current_user, chat_room_id: params[:room_id])
    ActionCable.server.broadcast("chat_room_#{params[:room_id]}", message: render_message(message))
  end

  private

  def render_message(message)
    ApplicationController.renderer.render(partial: 'messages/message', locals: { message: message })
  end
end
